--- makefile.unix	2010-09-15 17:10:14.161910034 +0000
+++ makefile.gentoo	2010-08-26 06:56:56.315435190 +0000
@@ -4,16 +4,11 @@
 
 
 INCLUDEPATHS= \
- -I"/usr/local/include/wx-2.9" \
- -I"/usr/local/lib/wx/include/gtk2-unicode-debug-static-2.9"
+ -I"/usr/include/wx-2.9" \
+ -I"/usr/lib/wx/include/gtk2-unicode-2.9" \
+ -I"/usr/include/db4.8"
 
-# for wxWidgets 2.9.1, add -l Xxf86vm
-WXLIBS= \
- -Wl,-Bstatic \
-   -l wx_gtk2ud-2.9 \
- -Wl,-Bdynamic \
-   -l gtk-x11-2.0 \
-   -l SM
+WXLIBS=$(shell /usr/bin/wx-config --libs)
 
 # for boost 1.37, add -mt to the boost libraries
 LIBS= \
@@ -46,7 +41,7 @@
     cryptopp/obj/cpu.o
 
 
-all: bitcoin
+all: bitcoin bitcoind
 
 
 obj/%.o: %.cpp $(HEADERS)
